libraries/standard/mqtt/src/mqtt.c

```diff
@@ -151,6 +169,61 @@ static MQTTStatus_t validateSubscribeUnsubscribeParams( const MQTTContext_t * co
 
 /*-----------------------------------------------------------*/
 
+static MQTTStatus_t receivePacket( MQTTContext_t * const pContext,
+                                   MQTTPacketInfo_t incomingPacket,
+                                   uint32_t remainingTimeMs )
+{
+    return MQTTSuccess;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t receiveConnack( MQTTContext_t * const pContext,
+                                    MQTTPacketInfo_t * const incomingPacket )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    uint8_t retryCount = 0U;
+
+    assert( pContext != NULL );
+    assert( incomingPacket != NULL );
+
+    do
+    {
+        /* Transport read for incoming CONNACK packet type and length. */
+        status = MQTT_GetIncomingPacketTypeAndLength( pContext->transportInterface.recv,
+                                                      pContext->transportInterface.networkContext,
+                                                      incomingPacket );
+
+        if( status == MQTTSuccess )
+        {
+            /* Reading the remainder of the packet by transport recv.
+             * The timeout is passed as 0 as it is only 2 more bytes to
+             * read for CONNACK. */
+            status = receivePacket( pContext,
+                                    *incomingPacket,
+                                    0U );
+        }
+
+        /* Loop until there is data to read or the retry count has not
+         * exceeded the limit. */
+    } while( ( status == MQTTNoDataAvailable ) &&
+             ( ++retryCount < CONNACK_RECV_RETRY_LIMIT ) );
```

Comment:
Are we sure we want to use a loop counter and not a timer? Based on how MQTT_Connect() is written, this will execute almost immediately after sending the CONNECT packet. What if the CONNACK is not available on the transport layer by that time?

Reply:
It may not be immediately executed because there is a timeout associated with a transport recv. So the sequence will be 
1. Send Connect packet using transport send.
2. Call transport recv. Read network buffer if data is available, otherwise timeout after the implementation configured transport recv timeout.
3. Keep calling the recv until the configured retry limit or a success or a failure to read data.
4. Read packet type and length.
5. Read the remainder of packet. for CONNACK it will be 2 bytes.

Reply:
Changed the logic to use timeouts. Please check